architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,550 @@
+
+    param application string
+    param tag string
+    param port int
+
+    resource container 'Applications.Core/containers@2023-10-01-preview' = {
+      name: 'container'
+      properties: {
+        application: application
+        container: {
+          image: 'ghcr.io/radius-project/samples/demo:${tag}'
+          ports: {
+            web: {
+              containerPort: port
+            }
+          }
+        }
+      }
+    }
+  parameters: |
+    application=''
+    tag='latest'
+    port=3000
+```
+
+### Bicep Module
+
+#### `main.bicep`
+```bicep
+param location string = resourceGroup().location
+param storageAccountName string
+
+// Import the storage module
+module storageModule 'storage.bicep' = {
+  name: 'storageModule'
+  params: {
+    location: location
+    storageAccountName: storageAccountName
+  }
+}
+
+// Output the storage account ID
+output storageAccountId string = storageModule.outputs.storageAccountId
+```
+
+#### `storage.bicep`
+```bicep
+// Define parameters
+param location string
+param storageAccountName string
+
+// Create a storage account
+resource storageAccount 'Microsoft.Storage/storageAccounts@2021-04-01' = {
+  name: storageAccountName
+  location: location
+  sku: {
+    name: 'Standard_LRS'
+  }
+  kind: 'StorageV2'
+  properties: {}
+}
+
+// Output the storage account ID
+output storageAccountId string = storageAccount.id
+```
+
+**Sample Output:**
+<!--
+Provide a sample output for the inputs provided above.
+-->
+```
+> rad bicep generate-kubernetes main.bicep --parameters storageAccountName=myaccount --outfile main.yaml
+
+Generating BicepDeployment resource...
+BicepDeployment resource generated at main.yaml
+
+To apply the BicepDeployment resource onto your cluster, run:
+kubectl apply -f main.yaml
+```
+
+#### `main.yaml`
+```yaml
+kind: BicepDeployment
+apiVersion: radapp.io/v1alpha3
+metadata:
+  name: env
+  namespace: radius-system
+spec:
+  template: |
+    param location string = resourceGroup().location
+    param storageAccountName string
+
+    // Import the storage module
+    module storageModule 'storage.bicep' = {
+      name: 'storageModule'
+      params: {
+        location: location
+        storageAccountName: storageAccountName
+      }
+    }
+
+    // Output the storage account ID
+    output storageAccountId string = storageModule.outputs.storageAccountId
+  modules:
+    - name: storageModule
+      template: |
+        param location string
+        param storageAccountName string
+
+        // Create a storage account
+        resource storageAccount 'Microsoft.Storage/storageAccounts@2021-04-01' = {
+          name: storageAccountName
+          location: location
+          sku: {
+            name: 'Standard_LRS'
+          }
+          kind: 'StorageV2'
+          properties: {}
+        }
+
+        // Output the storage account ID
+        output storageAccountId string = storageAccount.id
+  parameters: |
+    storageAccountName=myaccount
```

Comment:
This is a great example because it surfaces the complexity of moving the compilation to the server. 

Did you look into how the Bicep compiler handles this? What does the compiler output look like? Multiple files or a single file?